docs/BinaryProtocol.md

```diff
@@ -0,0 +1,565 @@
+ * `subType` → Subscription type: Exclusive, Shared, Failover
+ * `consumer_id` → Client generated consumer identifier. Needs to be unique
+    within the same connection
+ * `request_id` → Identifier for this request. Used to match the response with
+    the originating request. Needs to be unique within the same connection
+ * `consumer_name` → *(optional)* Clients can specify a consumer name. This
+    name can be used to track a particular consumer in the stats. Also, in
+    Failover subscription type, the name is used to decide which consumer is
+    elected as *master* (the one receiving messages): consumers are sorted by
+    their consumer name and the first one is elected master.
+
+##### Command Flow
+
+```json
+CommandFlow {
+  "consumer_id" : 1,
+  "messagePermits" : 1000
+}
+```
+
+Parameters:
+  * `consumer_id` → Id of an already established consumer
+  * `messagePermits` → Number of additional permits to grant to the broker for
+    pushing more messages
+
+##### Command Message
+
+Command `Message` is used by the broker to push messages to an existing consumer,
+within the limits of the given permits.
+
+This command is used in a frame that includes as well the message payload, for
+which the complete format is specified in the [payload commands](#payload-commands)
+section.
+
+```json
+CommandMessage {
+	"consumer_id" : 1,
+	"message_id" : {
+    "ledgerId" : 123,
+    "entryId" : 456
+  }
+}
+```
+
+
+#### Command Ack
+
+An `Ack` is used to signal to the broker that a given message has been
+successfully processed by the application and can be discarded by the broker.
+
+In addition, the broker will also maintain the consumer position based on the
+acknowledged messages.
+
+```json
+message CommandAck {
+	"consumer_id" : 1,
+	"ack_type" : "Individual",
+	"message_id" : {
+    "ledgerId" : 123,
+    "entryId" : 456
+  }
+}
+```
+
+Parameters:
+ * `consumer_id` → Id of an already established consumer
+ * `ack_type` → Type of acknowledgment: `Individual` or `Cumulative`
+ * `message_id` → Id of the message to acknowledge
+ * `validation_error` → *(optional)* Indicates that the consumer has discarded
+   the messages because of some problem like: `UncompressedSizeCorruption`,
+   `DecompressionError`, `ChecksumMismatch`, `BatchDeSerializeError`
+
+##### Command CloseConsumer
+
+***Note***: *This command can be sent by either producer or broker*.
+
+This command behaves the same as [`CloseProducer`](#command-closeproducer)
+
+##### Command RedeliverUnacknowledgedMessages
+
+A consumer can ask the broker to redeliver some or all the pending messages
+that were pushed to that particular consumer and not yet acknowledged.
+
+The protobuf object accepts a list of message ids that the consumer wants to
+be redelivered. If the list is empty, the broker will redeliver all the
+pending messages.
+
+With redelivery can be sent to the same consumer or spread across all available
+consumers in the case of a shared subscription.
```

Comment:
In case of Shared Subscription, redelivered messages can be sent to the same consumer or to all available consumers.